### PR TITLE
[WIP] MLM Training Objective

### PR DESCRIPTION
**Pull request by prady-saligram**
Introduces train_mlm.py, a new file adapted from train_lm.py, to support masked language modeling with dynamic masking as utilized in RoBERTa. A new class, MaskedLMDataset, has been implemented in text.py to handle dynamic masking. This class is instantiated and utilized within train_mlm.py, preserving all structural and sharding-related comments from the original train_lm.py to maintain clarity and continuity. The integration of MaskedLMDataset with the training script has been verified with appropriate parameters to ensure consistency with existing training workflows.

### CODE REVIEW

File: src/levanter/data/text.py [hunk 1]
```diff
@@ -64,6 +65,65 @@
...
+    def __iter__(self) -> Iterator[LmExample]:
+        key = self.key
+        sharding = jax.sharding.SingleDeviceSharding(jax.local_devices(backend="cpu")[0])
+
+        with use_cpu_device():
+            @functools.partial(eqx.filter_jit, out_shardings=sharding)
+            def _create_mlm_example(tokens, key):
+                tokens_array = tokens.array
+                
+                example = LmExample.causal(tokens=tokens, ignore_id=self.ignore_id)
```
Comment (dlwh) [inline src/levanter/data/text.py:102]: you need a non-causal attention mask for Roberta, and you need to set a loss_mask to be only the masked tokens

File: src/levanter/data/text.py [hunk 2]
```diff
@@ -64,6 +65,65 @@
...
+        QPos: Axis,
+        KPos: Axis,
+        mask_prob: float = 0.15,
+        key: Optional[PRNGKeyArray] = None,
+        ignore_index: Optional[int] = DEFAULT_IGNORE_INDEX,
+    ):
+        self.dataset = dataset
+        self.QPos = QPos
+        self.KPos = KPos
+        self.mask_prob = mask_prob
```
Comment (dlwh) [inline src/levanter/data/text.py:81]: should there be a mask_prob and a noise_prob?

File: src/levanter/data/text.py [hunk 3]
```diff
@@ -120,6 +180,8 @@ def _create_lm_example(tokens, key):
                 yield example
 
 
+
+
```
Comment (dlwh) [inline src/levanter/data/text.py:184]:
```suggestion
```

File: src/levanter/data/text.py (hunk 1 above)
Comment (dlwh) [reply src/levanter/data/text.py:102]: you also can't use the current LmExample actually because you need a separate targets field (with the non-masked tokens). With more work you could avoid the need for targets (with just masked tokens), but probably better to add an `targets: Optional[NamedArray]` to the class (or make your own class)
